### PR TITLE
[Timelock Partitioning] Part 4: Implement `PaxosBatch*` Apis.

### PR DESCRIPTION
**Pull request by felixdesouza**
**Goals (and why)**:
Implementing the apis defined in #4061.

**Implementation Description (bullets)**:

**Testing (What was existing testing like?  What have you done to improve it?)**:
Currently pending.

**Concerns (what feedback would you like?)**:

**Where should we start reviewing?**:
Anywhere

**Priority (whenever / two weeks / yesterday)**:

### CODE REVIEW

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/BatchPaxosAcceptorResource.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+    private void primeCache(Set<Client> clients) {
+        Set<WithSeq<Client>> latestSequences = KeyedStream.of(clients)
+                .map(paxosComponents::acceptor)
+                .map(PaxosAcceptor::getLatestSequencePreparedOrAccepted)
+                .map((client, latestSeq) -> WithSeq.of(latestSeq, client))
+                .values()
+                .collect(toSet());
+
+        acceptorCache.updateSequenceNumbers(latestSequences);
```
[the remark this answers is not in the document]
Comment (felixdesouza) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/BatchPaxosAcceptorResource.java]: will update `AcceptorCache` docs

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AcceptorCache.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+     *     <li>
+     *         If the {@code cacheKey} is invalid either because it was not issued by the cache or it has expired for
+     *         example, the implementation must throw a {@link InvalidAcceptorCacheKeyException}.
+     *     </li>
+     * </ul>
+     *
+     * @param cacheKey effective point in time where updates are requested from
+     * @return a {@link AcceptorCacheDigest digest} containing the updates and a new cache key if any
+     * @throws InvalidAcceptorCacheKeyException when {@code cacheKey} is missing or invalid
+     */
```
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AcceptorCache.java]: Semantics question (as far as I can tell either way is fine) - If the same cache key is provided twice, we should:

1. on the second time still return updates from that point on, assuming the key hasn't expired, but the next cache key should be different, regardless of the situation.
2. on the second time still return updates from that point on, assuming the key hasn't expired, but the next cache key should be the same, provided no further changes have been made.
3. invalidate the original key on the first request after answering, and then on the second time throw an `InvalidAcceptorCacheKeyException` 

I'd prefer (1) - (2) is complicated for what I think is little upside, and (3) might cause spurious cache failures/loads on a noisy network, but open to all.

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AcceptorCache.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+     *         {@link AcceptorCacheKey cache key} to use on the next invocation.
+     *     </li>
+     *     <li>
+     *         If the {@code cacheKey} is invalid either because it was not issued by the cache or it has expired for
+     *         example, the implementation must throw a {@link InvalidAcceptorCacheKeyException}.
+     *     </li>
+     * </ul>
+     *
+     * @param cacheKey effective point in time where updates are requested from
+     * @return a {@link AcceptorCacheDigest digest} containing the updates and a new cache key if any
```
[the remark this answers is not in the document]
Comment (felixdesouza) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AcceptorCache.java]: it reads as "a digest containing the updates" when rendered

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AcceptorCache.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (felixdesouza) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AcceptorCache.java]: 1 is what I've coded up

[the remark this answers is not in the document]
Comment (felixdesouza) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AcceptorCache.java]: 2 would limit amount of cache keys to number of followers. But my impl has a cache where keys are evicted if not accessed after 5 minutes and it's pretty easy/

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/BatchPaxosAcceptorResourceTests.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+                .<Client, WithSeq<PaxosProposalId>>builder()
+                .put(CLIENT_1, WithSeq.of(1, PROPOSAL_ID_1))
+                .putAll(CLIENT_2, WithSeq.of(1, PROPOSAL_ID_1), WithSeq.of(2, PROPOSAL_ID_2))
+                .build();
+
+        SetMultimap<Client, WithSeq<PaxosPromise>> expected = ImmutableSetMultimap
+                .<Client, WithSeq<PaxosPromise>>builder()
+                .put(CLIENT_1, WithSeq.of(1, paxosPromise(PROPOSAL_ID_1)))
+                .put(CLIENT_2, WithSeq.of(1, paxosPromise(PROPOSAL_ID_1)))
+                .put(CLIENT_2, WithSeq.of(2, paxosPromise(PROPOSAL_ID_2)))
```
[the remark this answers is not in the document]
Comment (felixdesouza) [reply timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/BatchPaxosAcceptorResourceTests.java]: my bad

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/BatchPaxosAcceptorResourceTests.java [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+    @Test
+    public void cachedEndpointDelegatesToCache() throws InvalidAcceptorCacheKeyException {
+        AcceptorCacheKey cacheKey = AcceptorCacheKey.of(UUID.randomUUID());
+        AcceptorCacheDigest digest = digest();
+        when(cache.updatesSinceCacheKey(cacheKey))
+                .thenReturn(Optional.of(digest));
+
+        assertThat(resource.latestSequencesPreparedOrAcceptedCached(cacheKey))
+                .contains(digest);
```
[the remark this answers is not in the document]
Comment (felixdesouza) [reply timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/BatchPaxosAcceptorResourceTests.java]: will do

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AcceptorCache.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (felixdesouza) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AcceptorCache.java]: I'm in agreement that it should be 1 if that wasn't clear. Did you want further clarification on the javadoc?

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/BatchPaxosAcceptorResourceTests.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (felixdesouza) [reply timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/BatchPaxosAcceptorResourceTests.java]: Actually no you don't, `when` is effectively doing that for me i.e. if I passed in a different cache key, I would not get the digest that I wanted.

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AcceptorCache.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AcceptorCache.java]: Ah, yeah that's right!

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AcceptorCache.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AcceptorCache.java]: Sure, would be nice to have. I wasn't sure if you'd coded the impl up yet, but wanted to check what you thought / how you were planning on handling this. Yep, let's stick with 1.

[the remark this answers is not in the document]
Comment (jeremyk-91) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/AcceptorCache.java]: Actually - no need to define what would happen in javadoc, but for the impl let's stick with 1

**Pull request merged**